cql3/statements/describe_statement.cc

```diff
@@ -0,0 +1,811 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+#include <exception>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <boost/range/adaptor/transformed.hpp>
+#include <boost/range/algorithm/sort.hpp>
+#include <boost/range/iterator_range_core.hpp>
+
+#include "cql3/column_specification.hh"
+#include "cql3/functions/function_name.hh"
+#include "cql3/statements/prepared_statement.hh"
+#include "exceptions/exceptions.hh"
+#include "keys.hh"
+#include "seastar/core/on_internal_error.hh"
+#include "service/client_state.hh"
+#include "types.hh"
+#include "cql3/query_processor.hh"
+#include "cql3/cql_statement.hh"
+#include "cql3/statements/raw/describe_statement.hh"
+#include "cql3/statements/describe_statement.hh"
+#include "seastar/core/shared_ptr.hh"
+#include "transport/messages/result_message.hh"
+#include "transport/messages/result_message_base.hh"
+#include "service/query_state.hh"
+#include "data_dictionary/keyspace_metadata.hh"
+#include "data_dictionary/data_dictionary.hh"
+#include "service/storage_proxy.hh"
+#include "cql3/ut_name.hh"
+#include "cql3/util.hh"
+#include "types/map.hh"
+#include "types/list.hh"
+#include "cql3/functions/functions.hh"
+#include "view_info.hh"
+#include "index/secondary_index_manager.hh"
+#include "cql3/functions/user_function.hh"
+#include "cql3/functions/user_aggregate.hh"
+#include "utils/overloaded_functor.hh"
+
+static logging::logger dlogger("describe");
+
+bool is_internal_keyspace(std::string_view name);
+
+namespace cql3 {
+
+namespace statements {
+
+namespace {
+
+struct description {
+    sstring _keyspace;
+    sstring _type;
+    sstring _name;
+    std::optional<sstring> _create_statement;
+
+    description(sstring keyspace, sstring type, sstring name, std::optional<sstring> create_statement)
+    : _keyspace(util::maybe_quote(std::move(keyspace)))
+    , _type(std::move(type))
+    , _name(util::maybe_quote(std::move(name)))
+    , _create_statement(std::move(create_statement)) {}
+
+    description(sstring keyspace, sstring type, sstring name)
+    : description(keyspace, type, name, std::nullopt) {}
+
+    void add_create_statement(sstring create_statement) {
+        _create_statement = std::move(create_statement);
+    }
+
+    std::vector<bytes_opt> serialize() const {
+        auto desc = std::vector<bytes_opt>{
+            {to_bytes(_keyspace)},
+            {to_bytes(_type)},
+            {to_bytes(_name)}
+        };
+
+        if (_create_statement) {
+            desc.push_back({to_bytes(*_create_statement)});
+        }
+
+        return desc;
+    }
+};
+
+template <typename D>
+concept Describable =
+    requires(D d, std::ostream& os) {
+        { d.describe(os) } -> std::same_as<std::ostream&>;
+    };
+
+template <Describable D>
+sstring get_create_statement(const D& element) {
+    std::ostringstream os;
+    element.describe(os);
+    return os.str();
```

Comment:
Alternative path is to have a describable interface and have those describable things inherit from it. Will see if it's worth it depending on template proliferation below.

Reply:
I've skipped it now to aviod refectoring loop, this PR is already taking too long